app/src/main/kotlin/com/simplemobiletools/calculator/helpers/CalculatorImpl.kt

```diff
@@ -3,214 +3,121 @@ package com.simplemobiletools.calculator.helpers
-        } else if (!sign.isEmpty()) {
-            setFormula(first + sign + second)
+    fun numpadClicked(id: Int) {
+        when (id) {
+            R.id.btn_decimal -> decimalClicked()
+            R.id.btn_0 -> addDigit(0)
+            R.id.btn_1 -> addDigit(1)
+            R.id.btn_2 -> addDigit(2)
+            R.id.btn_3 -> addDigit(3)
+            R.id.btn_4 -> addDigit(4)
+            R.id.btn_5 -> addDigit(5)
+            R.id.btn_6 -> addDigit(6)
+            R.id.btn_7 -> addDigit(7)
+            R.id.btn_8 -> addDigit(8)
+            R.id.btn_9 -> addDigit(9)
         }
     }
 
-    fun addDigit(number: Int) {
-        val currentValue = displayedNumber
-        val newValue = formatString(currentValue!! + number)
-        setValue(newValue)
-    }
-
-    private fun formatString(str: String): String {
-        // if the number contains a decimal, do not try removing the leading zero anymore, nor add group separator
-        // it would prevent writing values like 1.02
-        if (str.contains(".")) {
-            return str
+    fun handleOperation(operation: String) {
+        handleEquals()
+        operator = operation
+        if (OperationFactory.forId(operator!!, firstNumber, secondNumber) is UnaryOperation) {
+            handleEquals()
+        } else {
+            firstNumber += secondNumber
+            secondNumber = firstNumber - secondNumber
+            firstNumber -= secondNumber
+            decimalClicked = false
+            decimalCounter = 0
+            digits = 0
         }
-
-        val doubleValue = Formatter.stringToDouble(str)
-        return Formatter.doubleToString(doubleValue)
-    }
-
-    private fun updateResult(value: Double) {
-        setValue(Formatter.doubleToString(value))
-        mBaseValue = value
     }
 
-    private fun getDisplayedNumberAsDouble() = Formatter.stringToDouble(displayedNumber!!)
+    fun handleEquals() {
+        val operation = OperationFactory.forId(operator!!, firstNumber, secondNumber)
 
-    fun handleResult() {
-        mSecondValue = getDisplayedNumberAsDouble()
-        calculateResult()
-        mBaseValue = getDisplayedNumberAsDouble()
+        if (operation != null && (digits > 0 || operation is UnaryOperation)) {
+            resetValues()
+            firstNumber = operation.getResult()
+            setValue(Formatter.doubleToString(firstNumber))
+            setFormula(operation.getFormula())
+        }
     }
 
-    private fun handleRoot() {
-        mBaseValue = getDisplayedNumberAsDouble()
-        calculateResult()
+    fun handleReset() {
+        resetValues()
     }
 
-    private fun calculateResult() {
-        if (!mIsFirstOperation) {
-            updateFormula()
-        }
-
-        val operation = OperationFactory.forId(mLastOperation!!, mBaseValue, mSecondValue)
-
-        if (operation != null) {
-            updateResult(operation.getResult())
-        }
-
-        mIsFirstOperation = false
+    fun handleClear() {
```

Comment:
handleClear() is just calling handleReset() can we just remove this.

Reply:
no as the ui is still calling both handle clear and handle reset. The behavior will be update to have the "smart" clear functionality

Reply:
Ok I see